plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/CaseOfCase.hs

```diff
@@ -47,7 +75,24 @@ processTerm = \case
           mkIterApp
             ite
             [ cond
-            , (trueAnn, Delay trueAnn (Case ann true alts))
-            , (falseAnn, Delay falseAnn (Case ann false alts))
+            -- Here we call a single step of case-reduce in order to immediately clean up the
+            -- duplication of @alts@. Otherwise optimizing case-of-case-of-case-of... would create
+            -- exponential blowup of the case branches, which would eventually get deduplicated
+            -- with case-reduce, but only after that exponential blowup has already slowed the
+            -- optimizer down unnecessarily.
+            , (trueAnn, Delay trueAnn . CaseReduce.processTerm $ Case ann true alts)
+            , (falseAnn, Delay falseAnn . CaseReduce.processTerm $ Case ann false alts)
             ]
+  original@(Case annOuter (Case annInner scrut altsInner) altsOuter) ->
+    maybe
+      original
+      (Case annInner scrut)
+      (do
+        constrs <- for altsInner $ \case
+          c@(Constr _ i _)   -> Just (Left i, c)
+          c@(Constant _ val) -> Just (Right val, c)
+          _                  -> Nothing
+        -- See Note [Case-of-case and duplicating code].
+        guard $ length (nub . toList $ fmap fst constrs) == length constrs
+        pure $ constrs <&> \(_, c) -> CaseReduce.processTerm $ Case annOuter c altsOuter)
```

Comment:
I think this is overly restrictive. We should allow duplicates as long as they're of a small size. I'll implement that as a follow-up.